### ISSUE TITLE
Application crashes when trying to save some images

### ISSUE THREAD

**Issue by durga2112** [Astroua/CARTAvis#94]
Some of the files in the CARTAImages repository are causing crashes when I try to save a copy of them as an image. It doesn't take much to trigger this bug:

- start CARTA
- load either BigImagesTest/h_m51_b_s05_drz_sci.fits or TransposeTest/RaDecVel.fits into the image viewer
- attempt to save the image using the default settings

At this point, I get one of a few error messages. These are the messages I've seen so far:

- *** Error in `/home/jeff/scratch/build/dev/cpp/desktop/desktop': double free or corruption (out): 0x00007f2799b6c010 ***
- *** Error in `/home/jeff/scratch/build/dev/cpp/desktop/desktop': munmap_chunk(): invalid pointer: 0x00000000018cf830 ***
- *** Error in `/home/jeff/scratch/build/dev/cpp/desktop/desktop': free(): invalid next size (fast): 0x0000000002aeaf00 ***

It's possible that there are other files that can trigger this bug as well; I haven't tried all of them. I do find it curious that TransposeTest/VelRaDec.fits saves just fine without crashing, though. I am running code taken straight from the develop branch as of this writing.

**Comment by pfederl** [comment]
Where can I get access to these fits images?

**Comment by low-sky** [comment]
In the [CARTAImages](https://drive.google.com/a/ualberta.ca/folderview?id=0B71tg6U4W78KY3dXZ1JlYnlCSU0&usp=sharing) shared folder.  They are also part of the builds Alex is producing.

**Comment by pfederl** [comment]
Thanks Jeff.

I am trying to quickly look at this problem. I don't know what is causing it yet, but I have a feeling.

Anyhow, I noticed a different problem. About 1/10 times when I run this version of the viewer, I don't get the 'Save...' menu at all. @slovelan would you know why that is happening?

**Comment by astrilet** [comment]
I put those images into the google shared folder under "Images" directory

**Comment by slovelan** [comment]
I was able to reproduce the 'Save...' menu not appearing sometimes.  It 
appears to be a timing issue where sometimes
the variable needed to determine if the 'Save...' should be available 
has not been sent to the client when the menu
items are initialized.  There is a fix for it on the Susan branch.  If 
anyone needs the fix before a pull request is made,
please just copy DisplayWindowImage.js and DisplayWindowHistogram.js 
files from the Susan branch.

**Comment by pfederl** [comment]
Alex, where is 'the google shared folder'?

On Wed, Oct 21, 2015 at 9:10 AM, Alex Strilets <notifications@github.com>
wrote:

> I put those images into the google shared folder under "Images" directory
>
> On Tue, Oct 20, 2015 at 10:27 PM, Pavol Federl <notifications@github.com>
> wrote:
>
> > Thanks Jeff.
> >
> > I am trying to quickly look at this problem. I don't know what is causing
> > it yet, but I have a feeling.
> >
> > Anyhow, I noticed a different problem. About 1/10 times when I run this
> > version of the viewer, I don't get the 'Save...' menu at all. @slovelan
> > <https://github.com/slovelan> would you know why that is happening?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/Astroua/CARTAvis/issues/94#issuecomment-149778700>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Astroua/CARTAvis/issues/94#issuecomment-149926972>.
>



-- 
Pavol

**Comment by pfederl** [comment]
I uploaded a fix in my PavolHotfix branch: 731431d1606b97e1444743db2ff2678a4131022c

I believe the problem was caused by this line in ImageRenderService.h:

    void
    setInputView( NdArray::RawViewInterface::SharedPtr view, QString cacheId = QString(), int displayAxisX = 0, int displayAxisY= 0 );

Those are invalid default parameters, and since in ImageSaveService you don't set them, it creates all kinds of chaos for non-square images... It should be instead:

    void
    setInputView( NdArray::RawViewInterface::SharedPtr view, QString cacheId = QString(), int displayAxisX = 0, int displayAxisY= 1 );

On a related note, in ImageRenderService.cpp:

    template < class Pipeline >
    static void
    iView2qImage( NdArray::RawViewInterface * rawView, Pipeline & pipe, QImage & qImage, int displayAxisX, int displayAxisY )

this internal function will ONLY work with displayAxisX/Y set to 0,1 respectively. It will not work for any other values, and most likely lead to memory corruption. So I put in a fatal crash if someone tries to use it with values other than 0,1.

On yet another related subject... Your save image functionality behaves quite oddly once you actually interactively set the zoom level to something less than 1. At some point QImage of size 0x0 is created, which makes it impossible to paint to it, etc. I left in some debug statements that should allow you to debug it easier. The code will generate quite a few warning messages too.
